### DISCUSSION TITLE
Suggest new tags and identify similar tags based on embeddings

### DISCUSSION THREAD

**Discussion by Richie-Peak** [brianpetro/obsidian-smart-connections#274] [Ideas]
Hi! Love the plugin Brian, its the only AI plugin I actually ended up using for Obsidian.

A really cool potentially quite easy feature is smart tag suggestion. i.e. I could, for a given tag, get suggestions for notes in my vault that should have this tag but don't. 

How would this work? well, embedding tags is probably not a good idea as they are not descriptive. An easy first way might be to calculate the centroids of embeddings of all notes with a given tag. That gives you 'the average meaning' of a tag. Then it ranks notes by cosine distance to a given tag.

It could also use these centroids to calculate similarity between tags! It could suggest tags that are similar, even tags that might be merged

tbh, the same could probably be done for folders too. AI driven organisation of your vault!

**Comment by brianpetro** [comment]
Hi @Richie-Peak 

Thank you for your kind words and I'm glad to hear that you're finding the plugin useful.

I really like your idea of smart tag suggestions and AI-driven organization of the vault. It's definitely an interesting feature that could provide a lot of value to users.

Using the centroids of embeddings for tags and calculating the cosine distance to suggest notes is a great approach. Additionally, using the centroids to calculate similarity between tags and suggesting similar or mergeable tags is a clever use case.

Thank you for sharing your thoughts and suggestions! I do plan to implement something like this and it's helps me to receive feedback like this so I can better prioritize the timeline.

If you have any more ideas or feedback, please feel free to share. I appreciate your support!

Brian 🌴

**Comment by GoulartNogueira** [comment]
I like the idea.
I understand you need a **classifier**.

The embedding centroid approach is a very simple classifier:
- Given some samples (with a tag)
- Find the embeddings' centroid
- Evaluate the new embedding against the distance from the centroid.

It probably works for some cases, and it's certainly very lightweight.
But it might have some flaws. Mainly because you don't use any negative example to teach what should **not** have a tag.

So we should also consider other methods for classification, considering:
- accuracy
- training time
- run time

I suggest testing with:
- KNN (very easy to implement, has zero training time)
- Decision Tree (easy to implement, very fast run)
- SVM
